content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+
+#### hashes vs hashmaps vs dictionaries vs objects
+
+There is probably more to it than I know, but I find that this data structure has a different name depending on which language in which you're writing it:
+
+```javascript
+let previte = { name: 'Joe', age: 26 }
+```
+
+In the book (which uses Go), the author refers to it as a hash. In JavaScript, I would call this an object. In Python, one might call it a dictionary. Why can't we all decide on one name and use it across all languages? If someone knows the answer to this, please fill me in!
+
+#### first class functions
+
+I wrote a note to myself, "define this in your own words." It's when a programming language supports functions out of the box? No, it has to be more than that (don't most programming languages support first class functions?).
+
+Pulling from MDN [they say](https://developer.mozilla.org/en-US/docs/Glossary/First-class_Function),
+
+> A programming language is said to have **First-class functions** when functions in that language are treated like any other variable.
+
+Well, there you go! They're treated like any other variable. They elaborate and give three examples:
+
+1. function can be used as argument to another function
+2. function can be returned by another function
+3. function can be assigned as a value to a variable
+
+And now, our definition is complete.
+
+#### Thorsten was right
+
+Thorsten made a bet with us. He said,
+
+> I'd bet that you can follow this book along even if you've never written a single line of Go in your life.
+>
+> -Ball 10
+
+I tripped myself up in the beginning when initializing a Go project locally, but quickly solved that thanks to his help and friends on Twitter. So he was right! I don't think the language was an obstacle in anyway.
+
+> And if you, after reading this book, choose to write your own interpreter in another language this should come in handy.
+
+I wrote a note to remember this! I don't know if I'll do this, but it could be helpful for learning Reason or Rust more deeply. I'll leave this here as a reminder.
+
+### Chapter 1 - Lexing
+
+The first part of the interpreter: lexical analysis.
+
+#### lexical analysis, lexing, and lexer
+
+Lots of words that sound and look similar, but what do they all mean? "Lexical analysis" is taking the source code and transforming it into tokens. This process is also called "lexing." The "lexer" is the thing that does this. Thorsten mentions that it's also sometimes called tokenizer or scanner.
+
+https://excalidraw.com/#json=5086916491870208,sLUFfW5oMgrO3ApQwe4YQw
+
+> ...we're going to lex in our first step
+>
+> -Ball 15
+
+I wrote this down to highlight **lex** and remind myself how to use it in a sentence. An example, "The first thing our interpreter does is lex the source code to generate tokens. The tokens are then parsed into an abstract syntax tree."
```

Comment:
I'm honestly not sure whether using "lex" as a verb is correct :) What others are also using is "tokenize" or "scan".

Reply:
Whoops! Let me correct this